src/Microsoft.AspNetCore.Razor.OmniSharpPlugin/MSBuildProjectDocumentChangeDetector.cs

```diff
@@ -0,0 +1,143 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.IO;
+using System.Linq;
+using Microsoft.Build.Execution;
+using OmniSharp.MSBuild.Notification;
+
+namespace Microsoft.AspNetCore.Razor.OmniSharpPlugin
+{
+    [Shared]
+    [Export(typeof(IMSBuildEventSink))]
+    internal class MSBuildProjectDocumentChangeDetector : IMSBuildEventSink
+    {
+        private const string MSBuildProjectFullPathPropertyName = "MSBuildProjectFullPath";
+        private const string MSBuildProjectDirectoryPropertyName = "MSBuildProjectDirectory";
+        private static readonly IReadOnlyList<string> RazorFileExtensions = new[] { "razor", "cshtml" };
```

Comment:
Not asking to change but I feel like this should ideally come from Razor. We probably need this in other places too.